x/gamm/keeper/msg_server_test.go

```diff
@@ -356,3 +361,158 @@ func (suite *KeeperTestSuite) TestExitPool_Events() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestMsgMigrateShares() {
+	defaultAccount := suite.TestAccs[0]
+	defaultGammShares := sdk.NewCoin("gamm/pool/1", sdk.MustNewDecFromStr("100000000000000000000").RoundInt())
+	invalidGammShares := sdk.NewCoin("gamm/pool/1", sdk.MustNewDecFromStr("100000000000000000001").RoundInt())
+	defaultAccountFunds := sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(200000000000)), sdk.NewCoin("usdc", sdk.NewInt(200000000000)))
+
+	type param struct {
+		sender                sdk.AccAddress
+		sharesToMigrateDenom  string
+		sharesToMigrateAmount sdk.Int
+		poolIdEntering        uint64
+	}
+
+	tests := []struct {
+		name                       string
+		param                      param
+		expectedErr                error
+		sharesToCreate             sdk.Int
+		expectedMigrateShareEvents int
+		expectedMessageEvents      int
+	}{
+		{
+			name: "migrate all of the shares",
+			param: param{
+				sender:                defaultAccount,
+				sharesToMigrateDenom:  defaultGammShares.Denom,
+				sharesToMigrateAmount: defaultGammShares.Amount,
+				poolIdEntering:        2,
+			},
+			sharesToCreate:             defaultGammShares.Amount,
+			expectedMigrateShareEvents: 1,
+			expectedMessageEvents:      3, // 1 exitPool, 1 createPosition, 1 migrateShares.
+		},
+		{
+			name: "migrate half of the shares",
+			param: param{
+				sender:                defaultAccount,
+				sharesToMigrateDenom:  defaultGammShares.Denom,
+				sharesToMigrateAmount: defaultGammShares.Amount.Quo(sdk.NewInt(2)),
+				poolIdEntering:        2,
+			},
+			sharesToCreate:             defaultGammShares.Amount,
+			expectedMigrateShareEvents: 1,
+			expectedMessageEvents:      3, // 1 exitPool, 1 createPosition, 1 migrateShares.
+		},
+		{
+			name: "double the created shares, migrate 1/4 of the shares",
+			param: param{
+				sender:                defaultAccount,
+				sharesToMigrateDenom:  defaultGammShares.Denom,
+				sharesToMigrateAmount: defaultGammShares.Amount.Quo(sdk.NewInt(2)),
+				poolIdEntering:        2,
+			},
+			sharesToCreate:             defaultGammShares.Amount.Mul(sdk.NewInt(2)),
+			expectedMigrateShareEvents: 1,
+			expectedMessageEvents:      3, // 1 exitPool, 1 createPosition, 1 migrateShares.
+		},
+		{
+			name: "error: attempt to migrate shares from non-existent pool",
```

Comment:
we should have a test case trying to migrate to a pool that doesn't exist

Reply:
https://github.com/osmosis-labs/osmosis/pull/4051/commits/30e8c1f1bf5506b37ea07553537ea6b31c2a7f43